autoencoder.py

```diff
@@ -22,51 +37,132 @@ def __init__(self, intermediate_dim, code_dim, original_dim, num_hidden_layers,
+                          original)))
+
+
+def train(
+    loss,
+    model,
+    opt,
+    original,
+    ):
 
-def train(loss, model, opt, original):
     with tf.GradientTape() as tape:
         reconstructed = model(original)
         reconstruction_error = loss(reconstructed, original)
-    gradients = tape.gradient(reconstruction_error, model.trainable_variables)
+    gradients = tape.gradient(reconstruction_error,
+                              model.trainable_variables)
     gradient_variables = zip(gradients, model.trainable_variables)
     opt.apply_gradients(gradient_variables)
+
+    train_loss(reconstruction_error)
+
     return reconstruction_error
 
-def train_loop(model, loss, dataset, epochs):
-  opt = tf.optimizers.Adam(learning_rate=1e-3)
-  for epoch in range(epochs):
-      epoch_loss = 0
-      for batch_features in dataset:
-          loss_values = train(loss, model, opt, batch_features)
-          epoch_loss += loss_values
-      model.loss.append(tf.reduce_mean(epoch_loss))
-      print('Epoch {}/{}. Loss: {}'.format(epoch + 1, epochs, tf.reduce_mean(epoch_loss)))
\ No newline at end of file
+
+def train_loop(
+    model,
+    loss,
+    dataset,
+    epochs,
+    ):
+
+    opt = model.opt()
+    for epoch in range(epochs):
+        epoch_loss = 0
+        for batch_features in dataset:
+            loss_values = train(loss, model, opt, batch_features)
+            with train_summary_writer.as_default():
+                tf.summary.scalar('loss', train_loss.result(),
+                                  step=epoch)
+            epoch_loss += loss_values
+        model.loss.append(tf.reduce_mean(epoch_loss))
+        print('Epoch {}/{}. Loss: {}'.format(epoch + 1, epochs,
+              tf.reduce_mean(epoch_loss)))
+        train_loss.reset_states()
+
+
+def log_images(test_features, model):
+
+  # Sets up a timestamped log directory.
+    logdir = 'logs/train_data' \
+        + datetime.datetime.now().strftime('%Y%m%d-%H%M%S')
+
+  # Creates a file writer for the log directory.
+    file_writer = tf.summary.create_file_writer(logdir)
+
+    originals = []
+    reconstructed = []
+
+    number = 10  # Number of digits to display
+
+    for index in range(number):
+
+		# Save original images
+        originals.append(test_features[index].reshape(28, 28))
+
+		# Save reconstruction images
+        reconstructed.append(model(test_features)[index].numpy().reshape(28,
+                             28))
+
+    originals = np.array(originals)
+    reconstructed = np.array(reconstructed)
+
+    with file_writer.as_default():
+        images = np.reshape(originals, (-1, 28, 28, 1))
+        tf.summary.image('Original', images, max_outputs=10, step=0)
+
+    with file_writer.as_default():
+        images = np.reshape(reconstructed, (-1, 28, 28, 1))
+        tf.summary.image('Reconstructed', images, max_outputs=10,
+                         step=0)
+
+
+train_loss = tf.keras.metrics.Mean('train_loss', dtype=tf.float32)
```

Comment:
Why do we have these lines? [165-168]
Kindly use [this notebook](https://github.com/AFAgarap/autoencoders/blob/master/notebooks/0-autoencoder.ipynb) as a template.

Reply:
Lines 165-168 are for the TensorBoard since the logging of the **training performance**, **original** & **reconstructed images** are within the module. I think I asked if the TensorBoard logging should be in the module, and I assumed you stated that would be fine.

Reply:
Oh, sorry. My bad. Let's move it to the notebook. The module itself should only consist of the class, while the helper module would have the helper functions like `train` and `train_step`.